activation/nipost_test.go

```diff
@@ -807,6 +807,344 @@ func TestNIPSTBuilder_PoetUnstable(t *testing.T) {
+				&types.Post{
+					Indices: []byte{1, 2, 3},
+				}, &types.PostInfo{
+					Nonce: &nonce,
+				}, nil,
+			)
+
+			postService := NewMockpostService(ctrl)
+			postService.EXPECT().Client(sig.NodeID()).Return(postClient, nil).AnyTimes()
+
+			// successfully registered to 2 poets
+			err = nipost.AddPoetRegistration(db, sig.NodeID(), nipost.PoETRegistration{
+				ChallengeHash: challengeHash,
+				Address:       poetProverAddr,
+				RoundID:       "1",
+				RoundEnd:      time.Now().Add(10 * time.Second),
+			})
+			require.NoError(t, err)
+
+			err = nipost.AddPoetRegistration(db, sig.NodeID(), nipost.PoETRegistration{
+				ChallengeHash: challengeHash,
+				Address:       poetProverAddr2,
+				RoundID:       "1",
+				RoundEnd:      time.Now().Add(10 * time.Second),
+			})
+			require.NoError(t, err)
+
+			// successful post exec
+			nb, err := NewNIPostBuilder(
+				db,
+				postService,
+				zaptest.NewLogger(t),
+				PoetConfig{},
+				mclock,
+				nil,
+				WithPoetServices(poetProver),
+			)
+			require.NoError(t, err)
+
+			nipost, err := nb.BuildNIPost(context.Background(), sig, challengeHash,
+				&types.NIPostChallenge{PublishEpoch: publishEpoch})
+			require.NoError(t, err)
+			require.NotNil(t, nipost)
+		})
+
+	t.Run("completely changed poet service AFTER round started -> fail, too late to register again",
+		func(t *testing.T) {
+			const publishEpoch = 5
+			challenge := types.NIPostChallenge{
+				PublishEpoch: publishEpoch,
+			}
+
+			ctrl := gomock.NewController(t)
+			currLayer := types.EpochID(publishEpoch - 1).FirstLayer()
+			genesis := time.Now().Add(-time.Duration(currLayer) * layerDuration)
+
+			poetProver := NewMockPoetService(ctrl)
+			poetProver.EXPECT().Address().Return(poetProverAddr).AnyTimes()
+
+			mclock := NewMocklayerClock(ctrl)
+			mclock.EXPECT().LayerToTime(gomock.Any()).DoAndReturn(
+				func(got types.LayerID) time.Time {
+					return genesis.Add(layerDuration * time.Duration(got))
+				},
+			).AnyTimes()
+
+			db := localsql.InMemory()
+
+			require.NoError(t, nipost.AddChallenge(db, sig.NodeID(), &challenge))
+			challengeHash := wire.NIPostChallengeToWireV1(&challenge).Hash()
+
+			// successfully registered to removed poet
+			err = nipost.AddPoetRegistration(db, sig.NodeID(), nipost.PoETRegistration{
+				ChallengeHash: challengeHash,
+				Address:       poetProverAddr2,
+				RoundID:       "1",
+				RoundEnd:      time.Now().Add(10 * time.Second),
+			})
+
+			require.NoError(t, err)
+			postService := NewMockpostService(ctrl)
+
+			nb, err := NewNIPostBuilder(
+				localsql.InMemory(),
+				postService,
+				zaptest.NewLogger(t),
+				PoetConfig{},
+				mclock,
+				nil,
+				WithPoetServices(poetProver),
+			)
+			require.NoError(t, err)
+
+			nipost, err := nb.BuildNIPost(context.Background(), sig, types.RandomHash(),
+				&types.NIPostChallenge{PublishEpoch: currLayer.GetEpoch()})
+			require.ErrorIs(t, err, ErrATXChallengeExpired)
```

Comment:
I'm not sure it should return `ErrATXChallengeExpired`. This will delete all existing registrations and the nipost challenge from the DB: https://github.com/spacemeshos/go-spacemesh/blob/4bec6559dac61f9bd2e094329b513218156fb0a1/activation/activation.go#L437-L444
What if the user made a typo in the config? The deleted registrations are unrecoverable.

Reply:
We could return some other error, but my logic was to continue existing flow: if we do not have passing registrations to conditions, than it make sense to drop them. 
Anyway phase0 is short, so even if we would return some other error and would try to notify user by log, that he will lose epoch, he has to read logs (?) and revert configuration in time. 
 So, from my perspective, it doesn't make sense to keep registrations just in case if user would revert config before registration deadline.

Reply:
But in this test, the user changed the config (potentially making a mistake) during the poet round, which takes 2w. If we reset the nipost builder state, there is no way of restoring it.

Reply:
fixed